### PR TITLE
Restart pgf's latex instance after bad latex inputs.

### LINKED ISSUE #15968: pgf backend: getting stuck with "BrokenPipeError: [Errno 32] Broken pipe"
### Bug report

`matplotlib.backends.backend_pgf.LatexManager` communicates with a running tex process to get dimensions for text strings. Sometimes the child tex process will break and pgf backend is not able to save any new plots after that.

**Code for reproduction**

```python
import matplotlib as mpl
import matplotlib.pyplot as plt

plt.plot([1, 3, 2])
plt.title(r"foo \oops bar")

try:
    plt.savefig("test.pgf")
except ValueError as e:
    print(f"catched exception: {e}")

# start from scratch
plt.clf()

#assert mpl.rcParams["pgf.texsystem"] == "xelatex"; import os; os.system("ps aux | grep xetex")

plt.plot([1, 3, 2])
plt.title(r"foo $z$ bar")
# throws BrokenPipeError: [Errno 32] Broken pipe
plt.savefig("test.pgf")
```

**Actual outcome**

```
catched exception: Error processing 'foo \oops bar'
LaTeX Output:
! Undefined control sequence.
<argument> ...00}{14.400000}\selectfont foo \oops 
                                                  bar
<*> ...00000}{14.400000}\selectfont foo \oops bar}
                                                  
No pages of output.
Transcript written on texput.log.

Traceback (most recent call last):
  File "test.py", line 20, in <module>
    plt.savefig("test.pgf")
<snip>
  File "<snip>/lib/python3.7/site-packages/matplotlib/backends/backend_pgf.py", line 249, in _stdin_writeln
    self.latex_stdin_utf8.flush()
BrokenPipeError: [Errno 32] Broken pipe
```

**Expected outcome**

No exception, pgf file successfully saved.

**Matplotlib version**

  * Operating system: Linux, macOS
  * Matplotlib version: 3.1.1
  * Matplotlib backend (`print(matplotlib.get_backend())`):
  * Python version: 3.7
  * Jupyter version (if applicable):
  * Other libraries: texlive 2019

All packages are provided by nixpkgs.

### PR DESCRIPTION
**Pull request by anntzer**
... and also move the warning about pgf-to-stream not supporting raster
images to when actually outputting a raster image (otherwise the warning
is spurious, and affects e.g. test_tex_restart_after_error).

Closes #15968.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/tests/test_backend_pgf.py [hunk 1]
```diff
@@ -276,3 +277,14 @@ def test_pdf_pages_lualatex():
         pdf.savefig(fig)
 
         assert pdf.get_pagecount() == 2
+
+
+def test_tex_restart_after_error():
```
Comment (veprbl) [inline lib/matplotlib/tests/test_backend_pgf.py]: Probably need to add
```python
@needs_xelatex
@pytest.mark.backend('pgf')
```
?

Comment (anntzer) [reply lib/matplotlib/tests/test_backend_pgf.py]: I think only the first one (the backend will be autoswitched anyways), thanks.

**Comment by veprbl** [comment]
Thank you all for working on this!

**Comment by anntzer** [comment]
You're welcome, thanks for pointing out the issue :)

**Comment by QuLogic** [comment]
This is being backported in the above PR.

**Pull request merged**